include/eosio/vm/debug_info.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <cstdint>
+#include <cstddef>
+#include <algorithm>
+
+namespace eosio::vm {
+
+struct null_debug_info {
+   using builder = null_debug_info;
+   void on_code_start(const void* compiled_base, const void* wasm_code_start) {}
+   void on_function_start(const void* code_addr, const void* wasm_addr) {}
+   void on_instr_start(const void* code_addr, const void* wasm_addr) {}
+   void on_code_end(const void* code_addr, const void* wasm_addr) {}
+   void set(const null_debug_info&) {}
+   void relocate(const void*) {}
+};
+
+// Maps a contiguous region of code to offsets onto the code section of the original wasm.
+class profile_instr_map {
+   struct addr_entry {
+      uint32_t offset;
+      uint32_t wasm_addr;
+   };
+
+public:
+
+   struct builder {
+      void on_code_start(const void* compiled_base, const void* wasm_code_start) {
+         code_base = compiled_base;
+         wasm_base = wasm_code_start;
+      }
+      void on_function_start(const void* code_addr, const void* wasm_addr) {
+         data.push_back({
+            static_cast<std::uint32_t>(reinterpret_cast<const char*>(code_addr) - reinterpret_cast<const char*>(code_base)),
+            static_cast<std::uint32_t>(reinterpret_cast<const char*>(wasm_addr) - reinterpret_cast<const char*>(wasm_base))
+         });
+      }
+      void on_instr_start(const void* code_addr, const void* wasm_addr) {
+         data.push_back({
+            static_cast<std::uint32_t>(reinterpret_cast<const char*>(code_addr) - reinterpret_cast<const char*>(code_base)),
+            static_cast<std::uint32_t>(reinterpret_cast<const char*>(wasm_addr) - reinterpret_cast<const char*>(wasm_base))
+         });
+      }
+      void on_code_end(const void* code_addr, const void* wasm_addr) {
+         code_end = code_addr;
+      }
+
+      const void* code_base;
+      const void* wasm_base;
+      const void* code_end;
+      std::vector<addr_entry> data;
+   };
+
+   void set(builder&& b) {
+      data = std::move(b.data);
+      std::sort(data.begin(), data.end(), [](const addr_entry& lhs, const addr_entry& rhs){ return lhs.offset < rhs.offset; });
+      base_address = b.code_base;
+      code_size = reinterpret_cast<const char*>(b.code_end) - reinterpret_cast<const char*>(base_address);
+      offset_to_addr = data.data();
+      offset_to_addr_len = data.size();
+   }
+
+   profile_instr_map() = default;
```

Comment:
default constructor leaves attributes uninitialized. Safest to go ahead and provide defaults when they are declared below.